Improve good to know workout in React (#1505)
enkidevs/curriculum: javascript/react/good-to-know/immutability-helpers-in-react.md

---

# Immutability helpers in React

---
## Content

> When opting in for an immutable style of data management, a significant speedup can be achieved by easily implementing a fast `shouldComponentUpdate()` lifecycle method. To do this, you would need a method that aids with comparing *versions* of your component's `state`.

Reviewer: I think we should remove "easily" and "fast" from the first sentence. First one is subjective and second one unnecessary. 

> When opting in for an immutable style of data management, a significant speedup can be achieved by implementing a `shouldComponentUpdate()` lifecycle method.

Let's take a look at the following example of mutating data:

```javascript
myData.x.y.z = 10;
// or...
myData.a.b.push(13);
```

> Given the fact that **JavaScript** wasn't designed for dealing with immutable data, using one of the above methods does not provide a way of determining which data has changed since the previous copy. To help with this, **React** used to provide an immutability helper (has since become legacy) which is now replaced by the `immutability-helper` library.

Reviewer: ```suggestion
Given the fact that **JavaScript** wasn't designed for dealing with immutable data, using one of the above methods does not provide a way of determining which data has changed after the mutation. To help with this, **React** used to provide an immutability helper (has since become legacy) which is now replaced by the `immutability-helper` library.
```

> This library can be installed using the NPM package manager by using the following command:

Reviewer 2: This `new` library

> This new library can be installed using the NPM package manager by using the following command:

Reviewer: ```suggestion
This new library can be installed using the NPM package manager:
```

```shell

> $ npm install immutability-helper --save

Reviewer: ```suggestion
npm install immutability-helper --save
```

```

Now, with the help of the immutability helper we would write:

```javascript
import update from "immutability-helper";

const newData = update(myData, {
  x: { y: { z: { $set: 10 } } },
  // myData.x.y.z = 10;
  a: { b: { $push: [13 ] } }
  // myData.a.b.push(13);
});
```

Although this format might need some getting used to, it provides a much better approach for determining which data has changed [1].

> In terms of notation, keys that are preceded by a `$` are called **commands** (`{push: array}`, `{apply: function}`, etc.) and the data that is being mutated is called the **target**. 

Reviewer: ```suggestion
In terms of notation, keys that are preceded by a `$` are called **commands** (`{ $push: array }`, `{ $apply: function }`, etc.) and the data that is being mutated is called the **target**. 
```

Here is how you would perform a shallow merge using the `update()` method:

```javascript
var obj = { a: 5, b: 3 };
var newObj = update(obj, {
  $merge: { b: 6, c: 7 }
});
//newObj = {a: 5, b: 6, c:7}
```

> Other commands worth mentioning are `{$unshift: array}` and `{$splice: array of arrays}`, but you can see all the available methods in the Resources section of this insight.

Reviewer: ```suggestion
Other commands worth mentioning are `{ $unshift: array }` and `{ $splice: array of arrays }`, but you can see all the available methods in the Resources section of this insight.
```

> Other worth mentioning intuitive commands are `{$unshift: array}` and `{$splice: array of arrays}`, but you can see all the available methods in the Resources section of this insight.

Reviewer: Word order seems a bit unnatural here...maybe we should say (and remove "intuitive" cause it's subjective)?

> Other commands worth mentioning are...

* `{`
* `}`

---

> 

Reviewer 2: There shouldn't be any new line between `---` and `##`

Author: 👍 I'm in the process of fixing all the rule issues
